nixos/modules/services/monitoring/dd-agent/dd-agent.nix

```diff
@@ -188,50 +257,71 @@ in {
 
     users.extraGroups.datadog.gid = config.ids.gids.datadog;
 
-    systemd.services.dd-agent = {
-      description = "Datadog agent monitor";
-      path = [ pkgs."dd-agent" pkgs.python pkgs.sysstat pkgs.procps ];
-      wantedBy = [ "multi-user.target" ];
-      serviceConfig = {
-        ExecStart = "${pkgs.dd-agent}/bin/dd-agent foreground";
-        User = "datadog";
-        Group = "datadog";
-        Restart = "always";
-        RestartSec = 2;
-      };
-      restartTriggers = [ pkgs.dd-agent ddConf diskConfig networkConfig postgresqlConfig nginxConfig mongoConfig jmxConfig processConfig ];
-    };
+    systemd.services = let
+      makeService = attrs: recursiveUpdate {
+        path = [ cfg.package pkgs.python pkgs.sysstat pkgs.procps ];
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          User = "datadog";
+          Group = "datadog";
+          Restart = "always";
+          RestartSec = 2;
+        };
+        restartTriggers = [ cfg.package ddConf5 diskConfig networkConfig postgresqlConfig nginxConfig mongoConfig jmxConfig processConfig ];
+      } attrs;
+    in {
+      dd-agent = makeService ({
+        description = "Datadog agent monitor";
+      } // (if newDog then {
+          preStart = ''
+            chown datadog /etc/dd-agent
+            rm -f /etc/dd-agent/auth_token
+          '';
+          script = ''
+            export DD_API_KEY=$(head -n1 ${cfg.apiKeyFile})
+            exec ${cfg.package}/bin/agent start -c /etc/dd-agent/datadog.yaml
+          '';
+          serviceConfig.PermissionsStartOnly = true;
+        } else {
+          serviceConfig.ExecStart = "${cfg.package}/bin/dd-agent foreground";
+        }));
 
-    systemd.services.dogstatsd = {
-      description = "Datadog statsd";
-      path = [ pkgs."dd-agent" pkgs.python pkgs.procps ];
-      wantedBy = [ "multi-user.target" ];
-      serviceConfig = {
-        ExecStart = "${pkgs.dd-agent}/bin/dogstatsd start";
-        User = "datadog";
-        Group = "datadog";
-        Type = "forking";
-        PIDFile = "/tmp/dogstatsd.pid";
-        Restart = "always";
-        RestartSec = 2;
-      };
-      restartTriggers = [ pkgs.dd-agent ddConf diskConfig networkConfig postgresqlConfig nginxConfig mongoConfig jmxConfig processConfig ];
-    };
+      # dogstatsd is a separate process in dd-agent v5, but integrated
+      # into datadog-agent v6.
+      dogstatsd = mkIf (!newDog) (makeService {
+        description = "Datadog statsd";
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/dogstatsd";
+          Type = "forking";
+          PIDFile = "/tmp/dogstatsd.pid";
```

Comment:
`/tmp/` is a strange place for this. I hope statsd checks that no other user has created a malicious symlink there.

Reply:
There are established locations for this stuff indeed. `/tmp` is not the right location.

For my use case this kind of security is not important, but for some people it might be.